File: android/Reguerta/app/src/main/java/com/reguerta/user/data/notifications/ChainedNotificationRepository.kt
package com.reguerta.user.data.notifications

import com.reguerta.user.domain.notifications.NotificationEvent
import com.reguerta.user.domain.notifications.NotificationRepository
import com.reguerta.user.domain.access.Member

class ChainedNotificationRepository(
    private val primary: NotificationRepository,
    private val fallback: NotificationRepository,
) : NotificationRepository {
    override suspend fun getNotificationsFor(member: Member): List<NotificationEvent> {
        val primaryNotifications = primary.getNotificationsFor(member)
        return if (primaryNotifications.isNotEmpty()) {
            primaryNotifications
        } else {
            fallback.getNotificationsFor(member)
        }
    }

    override suspend fun getReadNotificationIds(memberId: String): Set<String> {
        val primaryReadIds = runCatching { primary.getReadNotificationIds(memberId) }.getOrDefault(emptySet())
        val fallbackReadIds = fallback.getReadNotificationIds(memberId)
        return primaryReadIds + fallbackReadIds
    }

    override suspend fun markNotificationsRead(
        memberId: String,
        notificationIds: Set<String>,
        readAtMillis: Long,
    ) {
        if (notificationIds.isEmpty()) return
        fallback.markNotificationsRead(memberId, notificationIds, readAtMillis)
        runCatching {
            primary.markNotificationsRead(memberId, notificationIds, readAtMillis)
        }
    }

    override suspend fun sendNotification(event: NotificationEvent): NotificationEvent {
        val fallbackSaved = fallback.sendNotification(event)
        val primarySaved = runCatching { primary.sendNotification(event) }.getOrNull()
        return primarySaved ?: fallbackSaved
    }
}

File: ios/Reguerta/Reguerta/Data/Notifications/ChainedNotificationRepository.swift
import Foundation

actor ChainedNotificationRepository<Primary: NotificationRepository, Fallback: NotificationRepository>:
    NotificationRepository {
    private let primary: Primary
    private let fallback: Fallback

    init(primary: Primary, fallback: Fallback) {
        self.primary = primary
        self.fallback = fallback
    }

    func notifications(visibleTo member: Member) async throws -> [NotificationEvent] {
        let primaryNotifications = try await primary.notifications(visibleTo: member)
        if !primaryNotifications.isEmpty {
            return primaryNotifications
        }
        return try await fallback.notifications(visibleTo: member)
    }

    func allNotifications() async throws -> [NotificationEvent] {
        let primaryNotifications = try await primary.allNotifications()
        if !primaryNotifications.isEmpty {
            return primaryNotifications
        }
        return try await fallback.allNotifications()
    }

    func readNotificationIds(memberId: String) async throws -> Set<String> {
        async let primaryReadIds = primary.readNotificationIds(memberId: memberId)
        async let fallbackReadIds = fallback.readNotificationIds(memberId: memberId)
        return try await primaryReadIds.union(fallbackReadIds)
    }

    func markNotificationsRead(memberId: String, notificationIds: [String], readAtMillis: Int64) async throws {
        guard !notificationIds.isEmpty else { return }
        try await fallback.markNotificationsRead(
            memberId: memberId,
            notificationIds: notificationIds,
            readAtMillis: readAtMillis
        )
        try await primary.markNotificationsRead(
            memberId: memberId,
            notificationIds: notificationIds,
            readAtMillis: readAtMillis
        )
    }

    func send(event: NotificationEvent) async throws -> NotificationEvent {
        _ = try await fallback.send(event: event)
        return try await primary.send(event: event)
    }
}
